File: source/vito.php
<?php
/*****************************************************************************/
// list of VITO data
/*****************************************************************************/
$frameReadings = [
//	'Gerät-Info'			=> { 'addr' => 0x00F8, 'bytes' => 2, 'divider' =>  1, 'unp' => 's' },
	'Temp-Aussen'			=> [ 'addr' => 0x0101, 'bytes' => 2, 'divider' => 10, 'unp' => 'i' ],
	'Temp-Luft-Ein'			=> [ 'addr' => 0x0103, 'bytes' => 2, 'divider' => 10, 'unp' => 'i' ],
	'Temp-Luft-Aus'			=> [ 'addr' => 0x0104, 'bytes' => 2, 'divider' => 10, 'unp' => 'i' ],
	'Temp-Hzg-VL'			=> [ 'addr' => 0x0105, 'bytes' => 2, 'divider' => 10, 'unp' => 'i' ],
	'Temp-Hzg-RL'			=> [ 'addr' => 0x0106, 'bytes' => 2, 'divider' => 10, 'unp' => 'i' ],
	'Temp-WW'				=> [ 'addr' => 0x010D, 'bytes' => 2, 'divider' => 10, 'unp' => 'i' ],
	'Status-Verdichter'		=> [ 'addr' => 0x0400, 'bytes' => 1, 'divider' =>  1, 'unp' => 'b' ],
	'Status-Pumpe-Hzg'		=> [ 'addr' => 0x040D, 'bytes' => 1, 'divider' =>  1, 'unp' => 'b' ],
	'Status-Pumpe-WW'		=> [ 'addr' => 0x0414, 'bytes' => 1, 'divider' =>  1, 'unp' => 'b' ],
#
#	'Temp-WW-Soll'			=> [ 'addr' => 0x0020, 'bytes' => 4, fmat => 10, unp => 'f<' },
	'Verdichter-BetriebStd'	=> [ 'addr' => 0x0580, 'bytes' => 4, 'divider' => 10, 'unp' => 'i' ],
	'Verdichter-NrStarts'	=> [ 'addr' => 0x0500, 'bytes' => 4, 'divider' =>  1, 'unp' => 'i' ],
	'WP-Uhrzeit'			=> [ 'addr' => 0x08E0, 'bytes' => 8, 'divider' =>  1, 'unp' => 's' ],
#
	'Strom-WP'				=> [ 'addr' => 0xFFFE, 'bytes' => 2, 'divider' =>  1, 'unp' => 'i' ],
	'Strom-H'				=> [ 'addr' => 0xFFFF, 'bytes' => 2, 'divider' =>  1, 'unp' => 'i' ],
	'Strom-Datum'			=> [ 'addr' => 0xFFFF, 'bytes' => 4, 'divider' =>  1, 'unp' => 's' ],
];

$readStatus = ['Temp-Aussen','Temp-Hzg-VL','Temp-Hzg-RL','Temp-WW',
				'Status-Verdichter','Status-Pumpe-Hzg','Status-Pumpe-WW','Strom-H','Strom-WP'];
/*****************************************************************************/
/*****************************************************************************/
function BuildSendCommand($fp, $cmd, $key, $value)
{
global $frameReadings;

	set_time_limit(5);
	// enter cmd mode: send 16 00 00 as long as the return byte is 06
	$i = 5;
	do {
		fwrite($fp, pack('H*', "160000"), 3);
		$char0 = ord( fgetc($fp) );
	} while ( $char0!=0x06 and (--$i)>0 );
	if ($i==0)
	{	// error occured
		fclose ($fp);
		die ("ERROR: Polling could not be stopped, received: $char0");
	}
	// so far so good. build here the string
	//check if $key exists
	if ( !isset($frameReadings[$key]) )
		die ("ERROR: BuildSendCommand: parameter $key not found!");
	$addr = $frameReadings[$key]['addr'];
	$byts = $frameReadings[$key]['bytes'];
	if ($cmd == "read")
	{
		$frame = chr(0x05) . chr(0x00) . chr(0x01) . chr($addr>>8) . chr($addr%256) . chr($byts);
	}
	elseif ($cmd == "write")
	{
		$frame = chr(5+$byts) . chr(0x00) . chr(0x02) . chr($addr>>8) . chr($addr%256) . chr($byts);// . ;
		$frame .= ParseWrValue($key, $value);	// done!
	//	echo "This command is not yet supported!\n";
	}
	else
	{
		echo "BuildSendCommand: undefined command '$cmd'!";
		return undef;
	}
	$out = chr(0x41) . $frame . chr(CRC8($frame));	// add cmdID as first and crc as last element
/**/
/*	$out = "";
	/* convert the array to binary data string
	foreach ($frame as &$value)
		$out .= chr($value);
*/
	// send command string
	fwrite($fp, $out, strlen($out));

	return strlen($out);
}

/*****************************************************************************/
/*****************************************************************************/
function CRC8($st)
{
	for ($i=0, $crc=0, $len=strlen($st); $len>0; $i++, $len--)
	{
		$crc += ord($st[$i]);
		//echo "byte $i: ".ord($st[$i]).", crc: ". $crc."<br>";
	}
	$crc %= 256;
	return ( $crc );	// only LSB is needed
}

/*****************************************************************************/
/*****************************************************************************/
function ReceiveData($fp)
{
	set_time_limit(5);

//	echo "Getting: ";
	for ($i = 0, $len = 5, $crc = 0, $in = "";
												$len>0;
														++$i, --$len)
	{	// read single characters until size (byte 2 of the received stream) is reached
		$char0 = ord( fgetc($fp) );	// take the numeric value of the received character
		//var_dump($char0);
		//echo "Answer byte nr. $i is: $char0,\t\thex: ". bin2hex(chr(($char0))) . "<br>"; // print one byte
//		echo bin2hex(chr(($char0))) . " "; // print one byte
		if ($i==0)
		{	// check if the first byte is 06
			if ( $char0==0x41 )
				$i = 1;
			elseif ( $char0!=0x06 ) {
				echo "First byte ".$char0.", hex: ". dechex($char0) . " unknown, expected '0x06'!\n";
				continue;
			}
		}
		elseif ( $i==2 )
		{	// check size byte
			$len = $char0+2;
			//echo "Data size is: $char0<br>";
		}
		$in .= chr($char0);	// build answer string here, except byte 0
		if ($i>1 and $len>1)
			$crc += $char0;	// calculate crc, except bytes 0, 1 and last one
	}
	$crc %= 256;
//	echo "<br>";
	//echo "CRC is: $crc\t";
	// check CRC error
	if ( $crc!=$char0 )	// last received byte
	{	echo "\nGot CRC error for the incoming string '$in'!\n"; return 0;}

	return $in;
}

/*****************************************************************************/
/*****************************************************************************/
function ParseRecData($rs, $key)
{
global $frameReadings;
	set_time_limit(5);

//	$res = unpack('I*', $rs);
//	var_dump($res); echo "<br>"; return;
	$res = unpack('H*', $rs);
//	var_dump($res[1]);	echo "<br>";
//	echo "$rs <br>";

	$regx = "~41([\da-f]{2})([\da-f]{2})01([\da-f]{4})([\da-f]{2})([\da-f]*)[\da-f]{2}~";
	
	$m = preg_match($regx, $res[1], $matches);
//	var_dump( $m );	echo "<br>";
//	var_dump( $matches );	echo "<br>";
	if ( !$m )
		echo "No match found, this data could not be parsed: " . $res[1] . "\n";

	// command "list" only woks on numerical arrays, starting with index 0
	list( , $len, $err, $ad, $byts, $data) = $matches;
//	var_dump( $data );	echo "<br>";

//	var_dump( $frameReadings[$key] ); 	echo " =key<br>";
	// to extract non-numerical array key values use the function 'extract()'
	$addr = "Original";	// needed to extract the original key values into the symbol table
	extract($frameReadings[$key]);
//	echo "\$a = $addr; \$b = $bytes; \$c = $divider; \$d = $unp<br>";
/* workaround for wrong bytes returned from ATTiny */
	if ( ($addr&0xfffe)==0xfffe )
		$byts = 2;

	if ( $unp=='i' )	// integer
	{
		if ( $byts==2 )		$ret = Parse2Bytes($data);
		elseif ( $byts==4 )	$ret = Parse4Bytes($data);
		// divide with the divider
		$ret /= $divider;
	}
	elseif ( $unp=='b' )	// binary
	{
		if ( $data=='00' )
			$ret = '0';
		else
			$ret = '1';
	}
	elseif ( $unp=='s' )	// string
	{
		$ret = $data;
	}

return $ret;
}

/*****************************************************************************/
/*****************************************************************************/
function Parse2Bytes($st)
{	// ABCD -> CDAB
	$a = intval($st[2].$st[3].$st[0].$st[1], 16);
	if ( $a & 0x8000 ) $a = $a - 0x10000;	// resolve sign
	return ( $a );
}

/*****************************************************************************/
/*****************************************************************************/
function Parse4Bytes($st)
{	// ABCDEFGH -> GHEFCDAB
	$a = intval($st[6].$st[7].$st[4].$st[5].$st[2].$st[3].$st[0].$st[1], 16);
	if ( $a & 0x80000000 ) $a = $a - 0x100000000;	// resolve sign
	return ( $a );
}

/*****************************************************************************/
/*****************************************************************************/
function ParseWrValue($key, $val)
{
global $frameReadings;
	set_time_limit(5);

	// to extract non-numerical array key values use the function 'extract()'
	$addr = "Original";	// needed to extract the original key values into the symbol table
	extract($frameReadings[$key]);
//	echo "\$addr = $addr; \$bytes = $bytes; \$divider = $divider; \$unp = $unp\n";

	if ( $unp=='i' )	// integer
	{
		// multiply by the divider
		$val *= $divider;
		if ( $byts==2 )		$ret = Build2Bytes((int)$data);
		elseif ( $byts==4 )	$ret = Build4Bytes((int)$data);
	}
	elseif ( $unp=='b' )	// binary
	{
		$ret = chr($val);
	}
	elseif ( $unp=='s' )	// string
	{
		$ret = BuildStringValue($val);
	}

return $ret;
}

/*****************************************************************************/
/*****************************************************************************/
function Build2Bytes($dt)
{
	return (chr($dt>>8) . chr($dt%256));
}

/*****************************************************************************/
/*****************************************************************************/
function Build4Bytes($dt)
{
	return Build2Bytes($dt>>16) . chr($dt%65536);
}

/*****************************************************************************/
/*****************************************************************************/
function BuildStringValue($dt)
{
	$dtlen = strlen($dt);
	$ret = '';
	for ($i=0; $i<$dtlen/2; $i++)
		$ret .= chr(hexdec($dt[2*$i])*16+hexdec($dt[2*$i+1]));
	return $ret;
}

/*****************************************************************************/
/*
utils:
	base_convert($bin,2,10); 
/*****************************************************************************/
?>

File: source/vito.cpp
/*****************************************************************************/
// list of VITO data
/*****************************************************************************/
#include "sys_cfg.h"
#include "vito.h"

const char temp_aussen[] 				PROGMEM = "Temp-Aussen";
const char temp_luft_ein[]				PROGMEM = "Temp-Luft-Ein";
const char temp_luft_aus[]				PROGMEM = "Temp-Luft-Aus";
const char temp_hzg_vl[]				PROGMEM = "Temp-Hzg-VL";
const char temp_hzg_rl[]				PROGMEM = "Temp-Hzg-RL";
const char temp_ww[]						PROGMEM = "Temp-WW";
const char temp_ww_soll[]				PROGMEM = "Temp-WW-Soll";
const char status_verdichter[]		PROGMEM = "Status-Verdichter";
const char status_pumpe_hzg[]			PROGMEM = "Status-Pumpe-Hzg";
const char status_pumpe_ww[]			PROGMEM = "Status-Pumpe-WW";
const char verdichter_betriebstd[]	PROGMEM = "Verdichter-BetriebStd";
const char verdichter_nrstarts[]		PROGMEM = "Verdichter-NrStarts";
const char strom_wp[]					PROGMEM = "Strom-WP";
const char strom_h[]						PROGMEM = "Strom-H";
const char strom_datum[]				PROGMEM = "Strom-Datum";
const char wp_uhrzeit[]					PROGMEM = "WP-Uhrzeit";
const char wp_betriebsart[]			PROGMEM = "Betriebsart";
const char wp_einmal[]					PROGMEM = "WW-Einmal";

const vito_param_t vito_params[] PROGMEM = {
//	'Geraet-Info'	{ 'addr' => 0x00F8, 'bytes' => 2, 'divider' =>  1, 'unp' => 's' },
{temp_aussen,				0x0101, 2, 10, 'i'},
{temp_luft_ein,			0x0103, 2, 10, 'i'},
{temp_luft_aus,			0x0104, 2, 10, 'i'},
{temp_hzg_vl,				0x0105, 2, 10, 'i'},
{temp_hzg_rl,				0x0106, 2, 10, 'i'},
{temp_ww,					0x010D, 2, 10, 'i'},
{temp_ww_soll,				0x6000, 2, 10, 'i'},
{status_verdichter,		0x0400, 1,  1, 'b'},
{status_pumpe_hzg,		0x040D, 1,  1, 'b'},
{status_pumpe_ww,			0x0414, 1,  1, 'b'},
//
{verdichter_betriebstd,	0x0580, 4, 10, 'i'},
{verdichter_nrstarts,	0x0500, 4,  1, 'i'},
{wp_uhrzeit,				0x08E0, 8,  1, 's'},
{wp_betriebsart,			0xB000, 1,  1, 'i'},
//
//{ww_einmal,					0xB020, 1,  1, 'b'},	*** address not correct ***
//
{strom_wp,					0xFFFE, 2,  1, 'i'},
{strom_h,					0xFFFF, 2,  1, 'i'},
{strom_datum,				0xFFFF, 4,  1, 's'},
//
};

const char * const read_params[] PROGMEM = {
	temp_aussen, temp_hzg_vl, temp_hzg_rl, temp_ww,
	status_verdichter, status_pumpe_hzg, status_pumpe_ww,
//	strom_h, strom_wp
};

#define VITO_REC_START      0
#define VITO_REC_TRAIN_LEN  1

int8_t rec_len, ind;
vito_param_t vito_param;  // currently accessed parameter
byte send_frame[100];	// output data frame, must be long for string formatted data (ex. date)
byte rec_frame[100];	// input data frame, must be long for string formatted data (ex. date)
char param_name[30];
char param_readings[PARAM_READINGS_SIZE];  // last read-out parameter values in string format

uint8_t ParseWrValue(uint8_t in);
uint8_t Parse2Bytes(void);
uint8_t Parse4Bytes(void);
uint8_t Build2Bytes(void);
uint8_t Build4Bytes(void);
uint8_t BuildStringValue(void);

/*****************************************************************************/
/*****************************************************************************/
char * Vito_GetParamName(byte in)
{
	strcpy_P(param_name, (const char *)pgm_read_word(&vito_params[in].name));
	return param_name;
}
/*****************************************************************************/
/*****************************************************************************/
void Vito_LoadParam(byte in)
{
	memcpy_P(&vito_param, &vito_params[in], sizeof(vito_param_t));
}
/*****************************************************************************/
/*****************************************************************************/
int8_t GetKeyIndex(const char * key)
{
//check if key exists
	for (byte i=0; i<sizeof(vito_params)/sizeof(vito_param_t); i++) {
		if ( key==(const char *)pgm_read_word(&vito_params[i].name) )
			return i;
	}
	return -1;
}
/*****************************************************************************/
/*****************************************************************************/
void Vito_ReceiveInit(void)
{
	ind = 0;
	rec_len = 10;
}
/*****************************************************************************/
/*****************************************************************************/
byte CRC8(void)
{
	byte crc = 0;
	byte i = send_frame[1]+1;
//  Serial.print(F("CRC8: len is ")); Serial.println(i,HEX);
	while ( i>0 )
		crc += send_frame[i--];
	return ( crc );
}
/*****************************************************************************/
/*****************************************************************************/
byte Vito_BuildCommand(byte in, char cmd)
{
#if _DEBUG_>1
	Serial.println(F("Vito_BuildCommand started..."));
#endif
	// variable "in" points to the array index
	Vito_LoadParam(in);
//	send_frame = {0x41, length, 0x00, command, adrr_high, addr_low, nr_bytes, CRC/data};	// CRC in read mode
	send_frame[0] = 0x41; // constant
	send_frame[1] = 0x05; // nr. bytes to send without crc
	send_frame[2] = 0x00; // constant

	if ( cmd == 'r') {
		// example: 41 05 00 01 00 F8 02 00
		send_frame[3] = 0x01;	// command "READ"
		send_frame[4] = vito_param.addr>>8;
		send_frame[5] = vito_param.addr%256;
		send_frame[6] = vito_param.len; // nr bytes of the parameter to write
	}
	else if (cmd == 'w')
	{
		send_frame[1] += vito_param.len;	// adjust nr bytes to send
		send_frame[3] = 0x02;	// command "WRITE"
		send_frame[4] = vito_param.addr>>8;
		send_frame[5] = vito_param.addr%256;
		send_frame[6] = vito_param.len;
		ParseWrValue(in);	// convert data to write
	}
	else
	{
#if _DEBUG_>0
        Serial.print("ERROR: Vito_SendCommand: undefined command "); Serial.println(cmd);
#endif
		return 0;
	}
	// add CRC as last element
	send_frame[send_frame[1]+2] = CRC8();
#if _DEBUG_>1
	Serial.println(F("Vito_BuildCommand done."));
#endif
	return (send_frame[1]+3);	// number of bytes to send
}
/*****************************************************************************/
/*****************************************************************************/
byte Vito_ReceiveData(byte char0)
{
// example stream: 06 41 07 01 01 55 25 02 07 01 8D
//	static char char0;
	if ( ind==VITO_REC_START )	// is first byte?
	{	// check if the first byte is 06
		if ( char0==0x06 ) {	// only next byte should be stored
			return 0;		// get next byte
		}
		else if ( char0!=0x41 ) {
			return 0;
		}
	} else if ( ind==VITO_REC_TRAIN_LEN ) { // is second byte?
		// set size value
		rec_len = char0+2;	// get remaining bytes to be received
	}
	rec_frame[ind++] = char0;	// store answer string here

	if ( (--rec_len)>0 )  return 0;
	// check CRC error after last received byte
	//if ( crc!=char0 )	Serial.println(F("ERROR: Vito_ReceiveData: Got CRC error for the incoming string!"));
	Vito_ReceiveInit();  // end of reception, re-initialize status
	return 1;
}
/*****************************************************************************/
// return number of parsed bytes
/*****************************************************************************/
byte Vito_ParseRecData(void)
{
// example received stream: [06] 41 07 01 01 55 25 02 07 01 8D

	// check error byte
	if ( rec_frame[2]==0x03 ) {
#if _DEBUG_>1
      Serial.println(F("ERROR: frame data error!"));
#endif
      return 0;
   }
	// check CRC
	uint8_t len = rec_frame[1]+2;
	uint8_t crc = 0;
	for (uint8_t i=1; i<len; i++)
		crc += rec_frame[i];
	if ( rec_frame[len]!=crc ) {
#if _DEBUG_>1
      Serial.println(F("ERROR: frame CRC error!"));
#endif
      return 0;
   }
	// check received parameters with the requested ones
	// swap high-low bytes for address
	uint16_t ad = rec_frame[4];
	ad = ad<<8;
	ad += rec_frame[5];
//        Serial.print(F("parsed address: ")); Serial.println(ad, HEX);
	if ( ad!=vito_param.addr ) {
#if _DEBUG_>1
      Serial.println(F("ERROR: frame address mismatch!"));
#endif
      return 0;
   }
	if ( rec_frame[6]!=vito_param.len ) {
#if _DEBUG_>1
      Serial.println(F("ERROR: frame parameter length mismatch!"));
#endif
      return 0;
   }

	if ( vito_param.typ=='i' )	// integer
	{
		if ( vito_param.len==2 )	return Parse2Bytes();	// divide with the divider
		else if ( vito_param.len==4 )	return Parse4Bytes();	// divide with the divider
	}
	else if ( vito_param.typ=='b' )	// binary
	{
		if ( rec_frame[7]==0x00 )   rec_frame[0] = '0';
		else			    rec_frame[0] = '1';
		rec_frame[1] = 0;
//  Serial.print(F("parsed byte: ")); Serial.println((char*)rec_frame);
		return 1;
	}
	else if ( vito_param.typ=='s' )	// string
	{
//todo		return ParseString();
		return 1;	// do nothing
	}
#if _DEBUG_>1
	Serial.print(F("ERROR: parsed unknown format!"));
#endif
	return 0;
}
/*****************************************************************************/
/*****************************************************************************/
byte Parse2Bytes(void)
{	// data is in little-endian format
	#define x ( *(int16_t*)(rec_frame+7) )
	char * cPtr = (char*)rec_frame;
	if ( x&0x8000 ) { // sign bit
		x = ~x + 1;	// two's complement on bit level == signed -> unsigned
		*cPtr++ = '-'; // print '-' sign
	}
	if ( vito_param.divid==10 )
		sprintf_P(cPtr, PSTR("%01u.%1u"), x/10, x%10);
	else
		sprintf_P(cPtr, PSTR("%02u"), x);
	//  Serial.println(F("parsed word: ")); Serial.println((char*)rec_frame);
	return 2;
}
/*****************************************************************************/
/*****************************************************************************/
uint8_t Parse4Bytes(void)
{	// data is in little-endian format
	#define x ( *(int32_t*)(rec_frame+7) )
	char * cPtr = (char*)rec_frame;
	if ( x&0x80000000 ) { // sign bit
		x = ~x + 1;	// two's complement on bit level == signed -> unsigned
		*cPtr++ = '-';
	}
	if ( vito_param.divid==10 )
		sprintf_P(cPtr, PSTR("%01lu.%1u"), x/10, x%10);
	else
		sprintf_P(cPtr, PSTR("%02lu"), x);
	//  Serial.println(F("parsed long: ")); Serial.println((char*)rec_frame);
	return 4;
}
/*****************************************************************************/
/*****************************************************************************/
uint8_t ParseWrValue(uint8_t in)
{
	Vito_LoadParam(in);
	
	if ( vito_param.typ=='i' )	// integer
	{
		// multiply by the divider
		if ( vito_param.len==2 )	return Build2Bytes();
		else if ( vito_param.len==4 )	return Build4Bytes();
	}
	else if ( vito_param.typ=='b' )	// binary
	{
		if ( send_frame[7]>0 )	send_frame[7] = 0x01;
		//else		send_frame[7] = 0x00;
		return 1;	// parsed one byte
	}
	else if ( vito_param.typ=='s' )	// string
	{
		return BuildStringValue();
	}
#if _DEBUG_>1
	Serial.print(F("ERROR: ParseWrValue: parameter has unknown format!"));
#endif
	return 0;
}
/*****************************************************************************/
/*****************************************************************************/
byte Build2Bytes(void)
{	// nothing to do, data should be available in little-endian format
	int16_t dt = *(int16_t*)&send_frame[7];
	return 2;	// written bytes
}
/*****************************************************************************/
/*****************************************************************************/
byte Build4Bytes(void)
{	// nothing to do, data should be available in little-endian format
	int32_t dt = *(int32_t*)&send_frame[7];
	return 4;	// written bytes
}
/*****************************************************************************/
/*****************************************************************************/
byte BuildStringValue(void)
{	// input is string, convert it to hex
	byte * dt = send_frame+7;
	byte dtlen = strlen((char*)dt);
/*	uint8_t ret = '';
	for (uint8_t i=0; i<dtlen/2; i++)
		send_frame[7+i] = atoi((const char*)dt[2*i])<<4 + atoi((const char*)dt[2*i+1]);
*/
	return dtlen;///2;
}

